### PR TITLE
feat(bake): handle bundle errors, re-assemble full client payloads, initial error modal

### PR DESCRIPTION
**Pull request by paperclover**
Resolves these TODO items:

- A file failing doesn't place a file watcher. This bug affects `bun build --watch` but no one has noticed???
- Any changes to server component manifests are not propagated
- When `Cmd+R` is pressed, it must not serve a stale bundle. The design for this to be fast is already present, but bundle re-assembly is not yet implemented.
- A build containing failing and passing files should not discard the failed files.
- Error Modal Prototype
- Deduplicate bundling errors on client + server files.
- Visual feedback in the terminal when successful updates happen
- https://github.com/oven-sh/bun/issues/14476

`bundle_v2`'s linker is no longer being run for the dev server as using it in reloads allows subtle mistakes regarding import/export names. The plan is to implement these checks at runtime, where it can be more robustly detected. All the while, we get faster rebuilds as those rebuilds do much less work now.

I am leaving out serialization of JS values since i want to get this merged to start doing CSS work.

Remaining tasks for this merge
- [x] Crash when failing twice in a row
- [ ] WebSocket wiring to show the error modal for real
- [x] WebSocket wiring to dismiss error modals
- [ ] Actually write the error modal prototype

### CODE REVIEW

**Comment by robobun** [comment]
@paperdave, your commit 3a6ce6eba005e17bb8b1c94bff363eb6bc78e69e is building: [`#4718`](https://buildkite.com/bun/bun/builds/4718)

File: src/bake/DevServer.zig [hunk 1]
```diff
@@ -391,76 +479,164 @@ fn onIncrementalVisualizerCorked(resp: *Response) void {
...
-        }
+    var result = server_request_callback.call(
+        global,
+        .undefined,
+        &.{
+            js_request,
+            meta,
+            route.module_name_string.get() orelse str: {
+                const js = bun.String.createUTF8(
+                    bun.path.relative(dev.cwd, route.entry_point),
```
Comment (dylan-conway) [inline src/bake/DevServer.zig:576]: should this be posix only?

File: src/bake/DevServer.zig [hunk 2]
```diff
@@ -718,42 +1030,67 @@ pub fn finalizeBundle(
...
-    defer dev.client_graph.affected_by_update = .{};
-    dev.server_graph.affected_by_update = try DynamicBitSetUnmanaged.initEmpty(linker.allocator, dev.server_graph.bundled_files.count());
-    defer dev.client_graph.affected_by_update = .{};
+    _ = chunk[1].content.css; // TODO: Index CSS files
 
-    ctx.server_seen_bit_set = try bun.bit_set.DynamicBitSetUnmanaged.initEmpty(linker.allocator, dev.server_graph.bundled_files.count());
+    dev.client_graph.affected_by_trace = try DynamicBitSetUnmanaged.initEmpty(bv2.graph.allocator, dev.client_graph.bundled_files.count());
+    defer dev.client_graph.affected_by_trace = .{};
+    dev.server_graph.affected_by_trace = try DynamicBitSetUnmanaged.initEmpty(bv2.graph.allocator, dev.server_graph.bundled_files.count());
+    defer dev.client_graph.affected_by_trace = .{};
```
Comment (dylan-conway) [inline src/bake/DevServer.zig:1047]: deinit?

File: src/bake/DevServer.zig [hunk 3]
```diff
@@ -621,29 +795,162 @@ fn theRealBundlingFunction(
...
+    assert(route.server_state == .loaded); // page is unfit to load
+
+    dev.graph_safety_lock.lock();
+    defer dev.graph_safety_lock.unlock();
+
+    // Prepare bitsets
+    var sfa_state = std.heap.stackFallback(65536, dev.allocator);
+
+    const sfa = sfa_state.get();
+    dev.server_graph.affected_by_trace = try DynamicBitSetUnmanaged.initEmpty(sfa, dev.server_graph.bundled_files.count());
```
Comment (dylan-conway) [inline src/bake/DevServer.zig:916]: Is this going to overwrite data in `affected_by_tree`? If not, the memory isn't reused, could we just remove `affected_by_tree` from the incremental graph struct?

File: src/bake/DevServer.zig [hunk 4]
```diff
@@ -917,12 +1146,81 @@ fn sendJavaScriptSource(code: []const u8, resp: *Response) void {
     resp.end(code, true); // TODO: You should never call res.end(huge buffer)
 }
 
+fn sendSerializedFailures(
+    dev: *DevServer,
+    resp: *Response,
+    failures: []const SerializedFailure,
+    kind: enum { runtime, bundler, evaluation },
```
Comment (dylan-conway) [inline src/bake/DevServer.zig]: what's the difference between runtime and evaluation failures?

File: src/bake/DevServer.zig [hunk 5]
```diff
@@ -1059,12 +1387,19 @@ pub fn IncrementalGraph(side: bake.Side) type {
             prev_dependency: EdgeIndex.Optional,
         };
 
-        /// An index into `bundled_files`, `stale_files`, `first_dep`, `first_import`, or `affected_by_update`
-        pub const FileIndex = bun.GenericIndex(u32, File);
+        /// An index into `bundled_files`, `stale_files`, `first_dep`, `first_import`, or `affected_by_trace`
+        /// Top bits cannot be relied on due to `SerializedFailure.Owner.Packed`
+        pub const FileIndex = bun.GenericIndex(u30, File);
+        pub const framework_entry_point_index = FileIndex.init(0);
+        pub const react_refresh_index = if (side == .client) FileIndex.init(1);
```
Comment (dylan-conway) [inline src/bake/DevServer.zig:1401]: I'm surprised `side` doesn't need `comptime`

File: src/bake/DevServer.zig [hunk 6]
```diff
@@ -2199,23 +2962,49 @@ pub fn reload(dev: *DevServer, reload_task: *HotReloadTask) bun.OOM!void {
...
+    if (reload_file_list.isVisible()) {
+        reload_file_list.log("Hot update hits {d} files", .{files.items.len});
+        for (files.items) |f| {
+            reload_file_list.log("- {s} (.{s})", .{ f.path, @tagName(f.graph) });
+        }
+    }
+
     dev.incremental_result.reset();
+    defer {
+        std.sort.pdq(
```
Comment (dylan-conway) [inline src/bake/DevServer.zig:2985]: why do we sort first?

File: src/js_printer.zig [hunk 7]
```diff
@@ -4940,15 +4940,8 @@ fn NewPrinter(
...
-                    var slice = p.writer.slice();
-                    const to_print: []const u8 = if (slice.len > 1024) slice[slice.len - 1024 ..] else slice;
-
-                    if (to_print.len > 0) {
-                        Output.panic("\n<r><red>voluntary crash<r> while printing:<r>\n{s}\n---This is a <b>bug<r>. Not your fault.\n", .{to_print});
-                    } else {
-                        Output.panic("\n<r><red>voluntary crash<r> while printing. This is a <b>bug<r>. Not your fault.\n", .{});
-                    }
+                else => |tag| {
+                    Output.panic("Unexpected tag in printStmt: .{s}", .{@tagName(tag)});
```
Comment (dylan-conway) [inline src/js_printer.zig:4944]: did you hit this panic?

Comment (paperclover) [reply src/js_printer.zig:4944]: yes

File: src/bun.zig [hunk 8]
```diff
@@ -3851,19 +3858,26 @@ pub fn WeakPtr(comptime T: type, comptime weakable_field: std.meta.FieldEnum(T))
 pub const DebugThreadLock = if (Environment.allow_assert)
     struct {
         owning_thread: ?std.Thread.Id = null,
+        locked_at: crash_handler.StoredTrace = crash_handler.StoredTrace.empty,
```
Comment (dylan-conway) [inline src/bun.zig:3861]: cool

File: src/crash_handler.zig [hunk 9]
```diff
@@ -1601,6 +1601,49 @@ pub fn dumpStackTrace(trace: std.builtin.StackTrace) void {
...
+    pub fn trace(stored: *StoredTrace) std.builtin.StackTrace {
+        return .{
+            .index = stored.index,
+            .instruction_addresses = &stored.data,
+        };
+    }
+
+    pub fn capture(begin: ?usize) StoredTrace {
+        var stored: StoredTrace = undefined;
+        var frame = stored.trace();
```
Comment (dylan-conway) [inline src/crash_handler.zig:1625]: this looks like it's using undefined memory for `stored.index`

File: src/bake/DevServer.zig (hunk 3 above)
Comment (paperclover) [reply src/bake/DevServer.zig:916]: the memory is not reused. i honestly dont want to do this refactor right now, but it should just be an argument passed to traceDependencies/traceImports

File: src/crash_handler.zig (hunk 9 above)
Comment (paperclover) [reply src/crash_handler.zig:1625]: .index is never branched on in that function so it is ok

**Review by dylan-conway — APPROVED** [review]

**Pull request merged**